blackjax/vi/fullrank_vi.py

```diff
@@ -0,0 +1,148 @@
+# Copyright 2020- The Blackjax Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Callable, NamedTuple, Tuple
+
+import math
+import jax
+from jax import lax
+import jax.numpy as jnp
+import jax.scipy as jsp
+from jax.flatten_util import ravel_pytree
+from optax import GradientTransformation, OptState
+
+from blackjax.types import PRNGKey, PyTree, Array
+
+__all__ = ["FullrankVIState", "FullrankVIInfo", "sample", "generate_fullrank_logdensity", "step"]
+
+
+def _real_vector_to_cholesky(X):
+    m = len(X)
+    n = int(.5 * (-1 + math.sqrt(1 + 8 * m)))
+    tril = X[:-n]
+    diag = jax.nn.softplus(X[-n:])
+    mat = jnp.zeros((n, n))
+    mat = mat.at[jnp.tril_indices(n, -1)].set(tril)
+    mat = mat + jnp.expand_dims(diag, -1) * jnp.identity(n)
+    return mat
+
+
+class FullrankVIState(NamedTuple):
+    mu: PyTree
+    L: Array # Flattened scale tril
+    opt_state: OptState
+
+
+class FullrankVIInfo(NamedTuple):
+    elbo: float
+
+
+def init(
+    position: PyTree,
+    optimizer: GradientTransformation,
+    *optimizer_args,
+    **optimizer_kwargs
+) -> FullrankVIState:
+    """Initialize the fullrank VI state."""
+    mu = jax.tree_map(jnp.zeros_like, position) # Is this a good init strategy?
```

Comment:
We should also allow random initialization for both mu and L, maybe allowing user to pass a callable that takes random_key and shape as input (for zeros and ones we can just ignore the random_key)

Reply:
Sounds good, shall we also make the changes to MFVI?

Reply:
I think we need to make a choice and let users initialise manually if they want something different.

Reply:
In fact, the init function here only serves as a "default initialization option", if the users want to specify their own strategy, they could replace the call to "fullrank_vi.init()" https://github.com/blackjax-devs/blackjax/pull/479/files#diff-8923a0a4ea42b4d3c2e1756e182a25f482809bc5b1c23a601423f5908f0f03e2R36 with their own function